### PR TITLE
Implement reduce keyword for SmoothL1Loss

### PR DESCRIPTION
**Pull request by zou3519**
As per #264. When reduce is False, SmoothL1Loss outputs a loss per element of the input tensor. When reduce is True (default), the current behavior is kept.

### Test Plan
`test/run_test.sh`
Added unit test for the reduce=False case. Also added a reference function to the reduce=True unit test.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -27,7 +27,8 @@
 from torch.nn.parallel._functions import Broadcast
 from common_nn import NNTestCase, ModuleTest, CriterionTest, TestBase, \
     module_tests, criterion_tests, TEST_CUDA, TEST_MULTIGPU, TEST_CUDNN, \
-    TEST_CUDNN_VERSION, nllloss_reference, nllloss2d_reference
+    TEST_CUDNN_VERSION, nllloss_reference, nllloss2d_reference, \
+    smoothl1loss_reference
```
Comment (apaszke) [inline test/test_nn.py]: it might be nicer to have a dict of reference functions that you import here, and then index by module name. We wouldn't need to modify the import list every time then

File: torch/lib/THCUNN/SmoothL1Criterion.cu [hunk 2]
```diff
@@ -26,23 +26,72 @@ struct smoothl1_functor
 };
 
 template <typename Dtype>
+struct smoothl1_update_functor
+{
+  smoothl1_update_functor() {}
+
+  __host__ __device__ void operator()(
```
Comment (apaszke) [inline torch/lib/THCUNN/SmoothL1Criterion.cu]: `__forceinline__`

File: torch/lib/THCUNN/SmoothL1Criterion.cu [hunk 3]
```diff
@@ -26,23 +26,72 @@ struct smoothl1_functor
 };
 
 template <typename Dtype>
+struct smoothl1_update_functor
```
Comment (apaszke) [inline torch/lib/THCUNN/SmoothL1Criterion.cu]: nit: can you please mention `no_reduce` in the name?

File: torch/lib/THCUNN/SmoothL1Criterion.cu [hunk 4]
```diff
@@ -26,23 +26,72 @@ struct smoothl1_functor
...
+template <typename Dtype>
 struct smoothl1_updateGradInput_functor
 {
   const Dtype norm;
+  const Dtype *gradOutput;
 
-  smoothl1_updateGradInput_functor(Dtype norm_)
-    : norm(norm_)
+  smoothl1_updateGradInput_functor(Dtype norm_, Dtype *gradOutput_)
+    : norm(norm_), gradOutput(gradOutput_)
```
Comment (apaszke) [inline torch/lib/THCUNN/SmoothL1Criterion.cu]: `gradOutput` can be loaded only once

File: torch/lib/THCUNN/generic/SmoothL1Criterion.cu [hunk 5]
```diff
@@ -46,24 +57,37 @@ void THNN_(SmoothL1Criterion_updateGradInput)(
...
   );
 
+  THCTensor_(resizeAs)(state, gradInput, input);
+
+  if (!reduce) {
+    THCUNN_check_nElement(state, gradOutput, input);
+    THC_pointwiseApply3(state, input, target, gradInput,
+                        smoothl1_updateGradInput_no_reduce_functor<real>());
+    THC_pointwiseApply2(state, gradOutput, gradInput,
+                        multiply_and_store_functor<real>());
```
Comment (apaszke) [inline torch/lib/THCUNN/generic/SmoothL1Criterion.cu]: `THCTensor_(cmul)`

**Review by apaszke — APPROVED** [review]
Looks good. Exactly the same comments as in the L1Loss PR.
